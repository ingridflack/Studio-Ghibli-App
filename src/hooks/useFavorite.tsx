import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";
import { toast } from "react-toastify";
import { IMovie } from "../types";

interface FavoriteProviderProps {
  children: ReactNode;
}

interface FavoriteContextData {
  favorites?: IMovie[];
  checkFavorite: (movie: IMovie) => boolean;
  toggleFavorite: (movie: IMovie) => void;
}

const FavoriteContext = createContext<FavoriteContextData>(
  {} as FavoriteContextData
);

const STORAGE_FAVORITE_KEY = "@Ghibli:favorites";

export function FavoriteProvider({
  children,
}: FavoriteProviderProps): JSX.Element {
  const [favorites, setFavorites] = useState<IMovie[] | undefined>();

  useEffect(() => {
    const storagedFavorites = localStorage.getItem(STORAGE_FAVORITE_KEY);
    const favorites = storagedFavorites ? JSON.parse(storagedFavorites) : [];

    setFavorites(favorites);
  }, []);

  useEffect(() => {
    if (favorites) {
      localStorage.setItem(STORAGE_FAVORITE_KEY, JSON.stringify(favorites));
    }
  }, [favorites]);

  const checkFavorite = (movie: IMovie): boolean =>
    !!favorites?.some((item) => item.id === movie.id);

  const addFavorite = (movie: IMovie) => {
    setFavorites((favorites) => (favorites ? [...favorites, movie] : []));

    toast.success("Movie added to favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const removeFavorite = (movie: IMovie) => {
    setFavorites((favorites) =>
      favorites?.filter((item) => item.id !== movie.id)
    );

    toast.success("Movie removed from favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const toggleFavorite = (movie: IMovie) => {
    if (checkFavorite(movie)) {
      removeFavorite(movie);
    } else {
      addFavorite(movie);
    }
  };

  return (
    <FavoriteContext.Provider
      value={{
        favorites,
        checkFavorite,
        toggleFavorite,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}

export function useFavorite() {
  const context = useContext(FavoriteContext);

  return context;
}
