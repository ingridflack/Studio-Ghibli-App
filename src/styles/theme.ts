const theme = {
  colors: {
    background: "#000",

    gray: {
      white: "#f6f6f6",
      platinum: "#e1e1e6",
      silver: "#a8a8b3",
      lightGray: "#7b8c98",
      dark: "#141414",
    },

    primary: "#e50914",
    secondary: "#b20710",
  },
  fontSize: {
    tiny: "0.65rem",
    small: "0.75rem",
    medium: "0.875rem",
    normal: "1rem",
    large: "1.5rem",
  },
} as const;

export type Theme = typeof theme;

export default theme;
