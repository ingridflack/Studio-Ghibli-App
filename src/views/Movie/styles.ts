import styled, { css } from "styled-components";

export const Header = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.gray.white};
  `}
`;

export const Title = styled.h2`
  ${({ theme }) => css`
    color: ${theme.colors.gray.white};
    font-weight: 400;
    font-size: 1.4rem;
  `}
`;

export const ButtonsContainer = styled.div`
  display: flex;
  margin-top: 1rem;
`;

export const RTBox = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    border: 1px solid ${theme.colors.primary};
    width: max-content;
    padding: 0.25rem 1rem;
    border-radius: 0.25rem;
    margin-right: 1rem;

    svg {
      font-size: ${theme.fontSize.large};
      color: ${theme.colors.primary};
      margin-right: 0.5rem;
    }
  `}
`;

export const FavoriteButton = styled.button<{ active: boolean }>`
  align-items: center;
  background-color: transparent;
  border: none;
  border: 1px solid ${({ theme }) => theme.colors.gray.dark};
  border-radius: 0.25rem;
  color: ${({ theme }) => theme.colors.gray.lightGray};
  display: flex;
  padding: 0.5rem;
  transition: color 200ms ease, border-color 200ms ease;

  &:hover {
    color: ${({ theme }) => theme.colors.primary};
    border-color: ${({ theme }) => theme.colors.primary};
  }

  > svg {
    margin-right: 0.5rem;
  }

  ${({ active }) => {
    return (
      active &&
      css`
        &,
        &:hover {
          color: ${({ theme }) => theme.colors.primary};
          border-color: ${({ theme }) => theme.colors.primary};
        }
      `
    );
  }}
`;

export const Content = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.gray.white};
  `}
`;

export const InfoTitle = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.gray.silver};
    display: flex;
    font-size: ${theme.fontSize.normal};
    margin: 1rem 0;
  `}
`;

export const Description = styled.p`
  ${({ theme }) => css`
    line-height: 1.25rem;
    font-size: ${theme.fontSize.medium};
  `}
`;

export const ProducerInfo = styled.div`
  ${({ theme }) => css`
    display: flex;
    margin-top: 1rem;

    span + span::before {
      content: "•";
      font-size: ${theme.fontSize.medium};
      margin: 0rem 0.75rem;

      @media screen and (max-width: 576px) {
        content: "";
        margin: 0;
      }
    }

    span {
      font-size: ${theme.fontSize.medium};
    }

    strong {
      color: ${theme.colors.gray.platinum};
      font-size: ${theme.fontSize.normal};
    }

    @media screen and (max-width: 576px) {
      flex-direction: column;
    }

    span + span {
      @media screen and (max-width: 576px) {
        margin-top: 0.5rem;
      }
    }
  `}
`;

export const Cast = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 1rem;

  @media screen and (max-width: 576px) {
    grid-template-columns: 1fr;
  }
`;

export const Person = styled.div`
  display: flex;
  align-items: center;
`;

export const PersonInfo = styled.div`
  ${({ theme }) => css`
    margin-left: 1rem;

    p {
      font-size: ${theme.fontSize.normal};
      display: flex;
      align-items: center;
    }

    span {
      font-size: ${theme.fontSize.medium};
    }

    span + span::before {
      content: "•";
      font-size: ${theme.fontSize.medium};
      margin: 0rem 0.5rem;
    }
  `}
`;

export const PersonImage = styled.img`
  background-color: white;
  border-radius: 50%;
  width: 4rem;
`;
