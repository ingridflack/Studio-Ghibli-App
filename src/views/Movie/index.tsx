import { MouseEvent } from "react";

import { MainContainer } from "../../styles/shared";
import { IMovie, IPerson } from "../../types";

import { AiOutlineArrowLeft } from "react-icons/ai";
import { SiRottentomatoes } from "react-icons/si";
import { BsHeart, BsHeartFill } from "react-icons/bs";

import * as S from "./styles";
import { useFavorite } from "../../hooks/useFavorite";

interface MoviesPageProps {
  movie: IMovie;
  people: any;
}

const MovieView = ({ movie, people }: MoviesPageProps) => {
  const { title, description, director, producer, release_date, rt_score } =
    movie;

  console.log(movie, title);

  const { checkFavorite, toggleFavorite } = useFavorite();

  const isFavorite = checkFavorite(movie);

  const handleFavorite = (e: MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();

    toggleFavorite(movie);
  };

  return (
    <MainContainer>
      <S.Header>
        <S.BackButton>
          <AiOutlineArrowLeft />
        </S.BackButton>

        <S.Title>{title}</S.Title>

        <S.ButtonsContainer>
          {!!rt_score && (
            <S.RTBox>
              <SiRottentomatoes /> {rt_score}%
            </S.RTBox>
          )}

          <S.FavoriteButton active={isFavorite} onClick={handleFavorite}>
            {isFavorite ? (
              <>
                <BsHeartFill /> Favorite
              </>
            ) : (
              <>
                <BsHeart /> Add to favorites
              </>
            )}
          </S.FavoriteButton>
        </S.ButtonsContainer>
      </S.Header>

      <S.Content>
        <S.InfoTitle>Description</S.InfoTitle>

        <S.Description>{description}</S.Description>

        <S.ProducerInfo>
          <span>
            <strong>Director: </strong>
            {director || "Unknown"}
          </span>
          <span>
            <strong>Producer: </strong>
            {producer || "Unknown"}
          </span>
          <span>
            <strong>Release date: </strong>
            {release_date || "Unknown"}
          </span>
        </S.ProducerInfo>

        <S.InfoTitle>Cast</S.InfoTitle>
        <S.Cast>
          {people.map(({ id, name, age, gender }: IPerson) => (
            <S.Person key={id}>
              <S.PersonImage src="/assets/user-placeholder.png" />

              <S.PersonInfo>
                <p>{name || "Unknown"}</p>
                <span>{(age && `Age: ${age}`) || "Unknown"}</span>
                <span>{gender || "Unknown"}</span>
              </S.PersonInfo>
            </S.Person>
          ))}
        </S.Cast>
      </S.Content>
    </MainContainer>
  );
};

export default MovieView;
