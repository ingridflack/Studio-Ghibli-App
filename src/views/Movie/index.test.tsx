import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Movie from ".";
import theme from "../../styles/theme";

const useRouter = jest.spyOn(require("next/router"), "useRouter");
const useFavorite = jest.spyOn(
  require("../../hooks/useFavorite"),
  "useFavorite"
);

const MOVIE_MOCK = {
  id: "234",
  title: "Title test",
  description: "Description test",
  director: "Director test",
  producer: "Producer test",
  movie_banner: "/assets/studio-ghibli-logo.webp",
  rt_score: "90",
  release_date: "1980",
  people: ["Person 1"],
  original_title: "Original title test",
};

const PEOPLE_MOCK = [
  { name: "Person 1", gender: "Female", age: "12", id: "1" },
  { name: "Person 2", gender: "Female", age: "13", id: "2" },
];

describe("Movie", () => {
  it("should match snapshot", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      checkFavorite: jest.fn().mockReturnValue(true),
    });
    useRouter.mockImplementation(() => ({
      pathname: "/movie",
    }));

    const { container } = render(
      <ThemeProvider theme={theme}>
        <Movie movie={MOVIE_MOCK} people={PEOPLE_MOCK} />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
