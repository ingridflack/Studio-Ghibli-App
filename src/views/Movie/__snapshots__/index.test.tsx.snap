// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should match snapshot 1`] = `
<div>
  <div
    class="sc-5ad07d5-0 bLYghj"
  >
    <div
      class="sc-c8ddd29-0 iFqajX"
    >
      <img
        alt="Studio Ghibli logo"
        class="sc-c8ddd29-1 bMwEkK"
        height="72"
        src="/assets/studio-ghibli-logo.webp"
        width="128"
      />
      <nav>
        <a
          class="sc-c8ddd29-2 fflUWz"
          href="/"
        >
          Home
        </a>
        <a
          class="sc-c8ddd29-2 fflUWz"
          href="/favorites"
        >
          Favorites
        </a>
      </nav>
    </div>
    <div
      class="sc-c0d3bbf4-0 gIUZqm"
    >
      <h2
        class="sc-c0d3bbf4-1 jBqMiD"
      >
        Title test
      </h2>
      <div
        class="sc-c0d3bbf4-2 kpXWtk"
      >
        <div
          class="sc-c0d3bbf4-3 bCPdBq"
        >
          <svg
            fill="currentColor"
            height="1em"
            role="img"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title />
            <path
              d="M5.866 0L4.335 1.262l2.082 1.8c-2.629-.989-4.842 1.4-5.012 2.338 1.384-.323 2.24-.422 3.344-.335-7.042 4.634-4.978 13.148-1.434 16.094 5.784 4.612 13.77 3.202 17.91-1.316C27.26 13.363 22.993.65 10.86 2.766c.107-1.17.633-1.503 1.243-1.602-.89-1.493-3.67-.734-4.556 1.374C7.52 2.602 5.866 0 5.866 0zM4.422 7.217H6.9c2.673 0 2.898.012 3.55.202 1.06.307 1.868.973 2.313 1.904.05.106.092.206.13.305l7.623.008.027 2.912-2.745-.024v7.549l-2.982-.016v-7.522l-2.127.016a2.92 2.92 0 0 1-1.056 1.134c-.287.176-.3.19-.254.264.127.2 2.125 3.642 2.125 3.659l-3.39.019-2.013-3.376c-.034-.047-.122-.068-.344-.084l-.297-.02.037 3.48-3.075-.038zm3.016 2.288l.024.338c.014.186.024.729.024 1.206v.867l.582-.025c.32-.013.695-.049.833-.078.694-.146 1.048-.478 1.087-1.018.027-.378-.063-.636-.303-.87-.318-.309-.761-.416-1.733-.418Z"
            />
          </svg>
           
          90
          %
        </div>
        <button
          class="sc-c0d3bbf4-4 kltpfA"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1.314C12.438-3.248 23.534 4.735 8 15-7.534 4.736 3.562-3.248 8 1.314z"
              fill-rule="evenodd"
            />
          </svg>
           Favorite
        </button>
      </div>
    </div>
    <div
      class="sc-c0d3bbf4-5 dEmzL"
    >
      <span
        class="sc-c0d3bbf4-6 jiKiFF"
      >
        Description
      </span>
      <p
        class="sc-c0d3bbf4-7 hrxycD"
      >
        Description test
      </p>
      <div
        class="sc-c0d3bbf4-8 dsjlIu"
      >
        <span>
          <strong>
            Director: 
          </strong>
          Director test
        </span>
        <span>
          <strong>
            Producer: 
          </strong>
          Producer test
        </span>
        <span>
          <strong>
            Release date: 
          </strong>
          1980
        </span>
      </div>
      <span
        class="sc-c0d3bbf4-6 jiKiFF"
      >
        Cast
      </span>
      <div
        class="sc-c0d3bbf4-9 eqnGbk"
      >
        <div
          class="sc-e5673028-0 fnXhTa"
        >
          <img
            alt="Person 1's photo"
            class="sc-e5673028-2 cdofYZ"
            height="4rem"
            src="/assets/user-placeholder.png"
            width="4rem"
          />
          <div
            class="sc-e5673028-1 gsSUVD"
          >
            <p>
              Person 1
            </p>
            <span>
              Age: 12
            </span>
            <span>
              Female
            </span>
          </div>
        </div>
        <div
          class="sc-e5673028-0 fnXhTa"
        >
          <img
            alt="Person 2's photo"
            class="sc-e5673028-2 cdofYZ"
            height="4rem"
            src="/assets/user-placeholder.png"
            width="4rem"
          />
          <div
            class="sc-e5673028-1 gsSUVD"
          >
            <p>
              Person 2
            </p>
            <span>
              Age: 13
            </span>
            <span>
              Female
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
